File: io/src/main/scala/edu/ucr/cs/bdlab/beast/io/shapefilev2/DBFReader.scala
/*
 * Copyright 2022 University of California, Riverside
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package edu.ucr.cs.bdlab.beast.io.shapefilev2

import edu.ucr.cs.bdlab.beast.util.IOUtil
import org.apache.spark.sql.types.{BooleanType, DateType, DoubleType, IntegerType, LongType, StringType, StructField, StructType, TimestampType}

import java.io.DataInputStream
import java.nio.charset.StandardCharsets
import java.sql.Timestamp
import java.util.GregorianCalendar

/**
 * Reads DBF entries
 * @param in an input stream to a DBF file initialized at the beginning
 */
class DBFReader(in: DataInputStream) extends Iterator[Array[Any]] with AutoCloseable {
  /** The header of the DBF file */
  val header: DBFHeader = DBFHelper.readHeader(in)

  /** A temporary array to read a single record */
  private val recordBytes: Array[Byte] = new Array[Byte](header.recordSize - 1)

  /** A flag that is raised when the end of file is reached */
  private var eofReached: Boolean = false

  /** The index of the record that will be returned next */
  var iCurrentRecord: Int = 0

  /** The index of the record to be returned next */
  var iNextRecord: Int = 0

  /** The record that will be returned next. null if end of file is reached */
  private var nextRecord: Array[Any] = prefetchNext

  lazy val schema: StructType = {
    StructType(header.fieldDescriptors.map(f => {
      val dataType = f.fieldType match {
        case DBFHelper.TypeString => StringType
        case DBFHelper.TypeNumeric | DBFHelper.TypeFloat if f.decimalCount == 0 && f.fieldLength < 9 => IntegerType
        case DBFHelper.TypeNumeric | DBFHelper.TypeFloat if f.decimalCount == 0 && f.fieldLength >= 9 => LongType
        case DBFHelper.TypeNumeric | DBFHelper.TypeFloat if f.decimalCount > 0 => DoubleType
        case DBFHelper.TypeBoolean => BooleanType
        case DBFHelper.TypeDate => DateType
        case DBFHelper.TypeDatetime => TimestampType
      }
      StructField(f.fieldName, dataType)
    }))
  }

  override def hasNext: Boolean = nextRecord != null

  private def prefetchNext: Array[Any] = {
    while (!eofReached) {
      iNextRecord += 1
      val marker = in.readByte()
      if (marker == DBFHelper.EOFMarker) {
        eofReached = true
      } else if (marker == DBFHelper.ValidRecordMarker) {
        // A valid record. Read it
        in.readFully(recordBytes)
        return parseRecord(recordBytes)
      } else if (marker == DBFHelper.DeletedRecordMarker) {
        // A deleted record. Skip over it
        in.skipBytes(recordBytes.length)
      } else {
        throw new RuntimeException(s"Error parsing DBF file. Invalid marker ${marker}")
      }
    }
    // EOF reached
    null
  }

  override def next(): Array[Any] = {
    iCurrentRecord = iNextRecord
    val currentRecord = nextRecord
    nextRecord = prefetchNext
    currentRecord
  }

  /**
   * Parse bytes of a record into an array of values
   * @param bytes data read from the file
   * @return an array of parsed values according to the header
   */
  private def parseRecord(bytes: Array[Byte]): Array[Any] = {
    val values = new Array[Any](header.fieldDescriptors.length)
    var startOffset: Int = 0
    for (iField <- header.fieldDescriptors.indices) {
      val field = header.fieldDescriptors(iField)
      var endOffsetInclusive: Int = (bytes.length - 1) min (startOffset + field.fieldLength - 1)
      values(iField) = field.fieldType match {
        case DBFHelper.TypeString =>
          // String. Skip leading spaces or null values
          while (endOffsetInclusive >= startOffset &&
            (bytes(endOffsetInclusive) == ' ' || bytes(endOffsetInclusive) == '\0'))
            endOffsetInclusive -= 1
          if (startOffset > endOffsetInclusive)
            null
          else
            new String(bytes, startOffset, endOffsetInclusive - startOffset + 1, StandardCharsets.UTF_8)
        case DBFHelper.TypeNumeric | DBFHelper.TypeFloat =>
          // Skip any leading blanks (or non-digits)
          var adjustedStartOffset = startOffset
          while (adjustedStartOffset <= endOffsetInclusive &&
              !(bytes(adjustedStartOffset) >= '0' && bytes(adjustedStartOffset) <= '9') &&
              bytes(adjustedStartOffset) != '.' && bytes(adjustedStartOffset) != '-')
            adjustedStartOffset += 1
          // Check if it is an empty value
          if (adjustedStartOffset > endOffsetInclusive)
            null
          else {
            val strValue = new String(bytes, adjustedStartOffset,
              endOffsetInclusive - adjustedStartOffset + 1, StandardCharsets.US_ASCII)
            if (field.decimalCount == 0 && field.fieldLength < 10) {
              // Return value as integer
              strValue.toInt
            } else if (field.decimalCount == 0 && field.fieldLength < 19) {
              // Return value as long
              strValue.toLong
            } else {
              strValue.toDouble
            }
          }
        case DBFHelper.TypeDouble =>
          val v = IOUtil.readDoubleLittleEndian(bytes, startOffset)
          if (v.isNaN) null else v
        case DBFHelper.TypeDate =>
          val yyyy = (bytes(startOffset) - '0') * 1000 + (bytes(startOffset + 1) - '0') * 100 +
            (bytes(startOffset + 2) - '0') * 10 + (bytes(startOffset + 3) - '0')
          val mm = (bytes(startOffset + 4) - '0') * 10 + (bytes(startOffset + 5) - '0')
          val dd = (bytes(startOffset + 6) - '0') * 10 + (bytes(startOffset + 7) - '0')
          val date = new GregorianCalendar(DBFHelper.UTC)
          date.clear() // Important to clear the time
          date.set(yyyy, mm - 1, dd)
          new java.sql.Date(date.getTimeInMillis)
        case DBFHelper.TypeDatetime =>
          // 8 bytes - two longs, first for date, second for time.
          // The date is the number of days since  01/01/4713 BC.
          // Time is hours * 3600000L + minutes * 60000L + Seconds * 1000L
          val datepart = IOUtil.readIntLittleEndian(bytes, startOffset)
          val timepart = IOUtil.readIntLittleEndian(bytes, startOffset + 4)
          if (datepart == 0 && timepart == 0)
            null
          else
            new Timestamp(datepart * DBFHelper.MillisInOneDay + DBFHelper.DBFEpoch.getTimeInMillis + timepart)
        case DBFHelper.TypeBlockNumber =>
          // 10 digits representing a .DBT block number. The number is stored as a string,
          // right justified and padded with blanks.
          // Skip leading spaces
          while (startOffset <= endOffsetInclusive && bytes(startOffset) == ' ')
            startOffset += 1
          if (startOffset == endOffsetInclusive)
            null
          else
            new String(bytes, startOffset, endOffsetInclusive - startOffset + 1, StandardCharsets.UTF_8)
        case DBFHelper.TypeBoolean =>
          // Logical (Boolean) type
          // initialized to 0x20 (space) otherwise T or F
          bytes(startOffset) match {
            case ' ' => null
            case 'T' => true
            case 'F' => false
            case _ => null
          }
        case other => throw new RuntimeException(s"Field type '$other' not supported")
      }
      startOffset += field.fieldLength
    }
    values
  }

  override def close(): Unit = in.close()
}

File: io/src/main/java/edu/ucr/cs/bdlab/beast/io/shapefile/DBFReader.java
/*
 * Copyright 2018 University of California, Riverside
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package edu.ucr.cs.bdlab.beast.io.shapefile;

import edu.ucr.cs.bdlab.beast.common.BeastOptions;
import edu.ucr.cs.bdlab.beast.geolite.Feature;
import edu.ucr.cs.bdlab.beast.geolite.GeometryHelper;
import edu.ucr.cs.bdlab.beast.geolite.IFeature;
import edu.ucr.cs.bdlab.beast.util.IOUtil;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.spark.sql.types.*;

import java.io.DataInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.GregorianCalendar;
import java.util.List;

/**
 * Reads a DBase file as a sequence of features.
 * This class follows the file specifications at http://www.dbase.com/Knowledgebase/INT/db7_file_fmt.htm
 * and http://www.dbf2002.com/dbf-file-format.html
 */
public class DBFReader implements AutoCloseable {

  /**The header of the current file*/
  protected DBFHeader header;

  /**Total size of a record in bytes*/
  protected int recordSize;

  /**An array to hold the values of all fields of one record*/
  protected byte[] recordValue;

  /**The value currently pointed by this reader*/
  protected Feature feature;

  /**The input stream to the DBF file*/
  protected DataInputStream in;

  /**This flag is raised when the EOF marker is reached*/
  protected boolean eofReached;

  /**The index of the current record one-based. Used to calculate the progress in an easy way.*/
  protected int iRecord;

  /**Cache the path to the input file to report the error*/
  protected Path dbfPath;

  /**Field names and types as appear in the header*/
  protected StructField[] fields;

  /**The schema of the created object*/
  protected StructType schema;

  public void initialize(Path dbfPath, BeastOptions conf) throws IOException {
    FileSystem fs = dbfPath.getFileSystem(conf.loadIntoHadoopConf(null));
    initialize(fs.open(dbfPath), conf);
    // Cache the file path to use in reporting errors
    this.dbfPath = dbfPath;
  }

  public void initialize(DataInputStream in, BeastOptions conf) throws IOException {
    this.dbfPath = null; // Reset file name if it is initialized from a stream
    this.in = in;
    this.header = new DBFHeader();
    readHeader(this.in, header);
    readFieldDescriptors(this.in, header);
    if ((header.version & 0x7) == 4) {
      readFieldProperties(this.in, header);
    }
    iRecord = 0;
    eofReached = false;
    this.recordValue = new byte[this.recordSize];
    // Initialize field names and types to use for creating features
    this.fields = new StructField[1 + header.fieldDescriptors.length];
    this.fields[0] = new StructField("g", GeometryHelper.GeometryType, true, Metadata.empty());
    for (int iField = 0; iField < header.fieldDescriptors.length; iField++) {
      FieldDescriptor fieldDescriptor = header.fieldDescriptors[iField];
      String fieldName = fieldDescriptor.getFieldName();
      DataType fieldType;
      switch (fieldDescriptor.fieldType) {
        case DBFConstants.TypeString:
        case DBFConstants.TypeBlockNumber:
          fieldType = DataTypes.StringType;
          break;
        case DBFConstants.TypeNumeric:
        case DBFConstants.TypeFloat:
          if (fieldDescriptor.decimalCount > 0)
            fieldType = DataTypes.DoubleType;
          else if (fieldDescriptor.fieldLength - fieldDescriptor.decimalCount <= 9)
            fieldType = DataTypes.IntegerType;
          else
            fieldType = DataTypes.LongType;
          break;
        case DBFConstants.TypeDouble:
          fieldType = DataTypes.DoubleType;
          break;
        case DBFConstants.TypeDate:
          fieldType = DataTypes.DateType;
          break;
        case DBFConstants.TypeDatetime:
          fieldType = DataTypes.TimestampType;
          break;
        case DBFConstants.TypeBoolean:
          fieldType = DataTypes.BooleanType;
          break;
        default:
          throw new RuntimeException("Unrecognized type "+fieldDescriptor.fieldType);
      }
      fields[1 + iField] = new StructField(fieldName, fieldType, true, Metadata.empty());
    }
    this.schema = new StructType(fields);
  }

  private void readFieldProperties(DataInputStream in, DBFHeader header) {
    throw new RuntimeException("Not supported yet");
  }

  /**
   * Read the first part of the header before the field descriptors
   * @param in the input stream to read from
   * @param header the header of the DBF file
   * @throws IOException if an error happens while reading the header
   */
  protected void readHeader(DataInputStream in, DBFHeader header) throws IOException {
    header.version = in.readByte();
    header.dateLastUpdatedYY = (short) in.readUnsignedByte();
    header.dateLastUpdatedMM = (short) in.readUnsignedByte();
    header.dateLastUpdatedDD = (short) in.readUnsignedByte();
    header.numRecords = IOUtil.readIntLittleEndian(in);
    header.headerSize = IOUtil.readShortLittleEndian(in);
    header.recordSize = IOUtil.readShortLittleEndian(in);
    in.skipBytes(2); // Reserved; filled with zeros.
    header.incomplete = in.readByte();
    header.encryption = in.readByte();
    in.skipBytes(12); // Reserved for multi-user processing.
    header.mdxFlag = in.readByte();
    header.driverID = in.readByte();
    in.skipBytes(2); // Reserved; filled with zeros.
    if ((header.version & 0x7) == 4) {
      // dBase version 7
      in.read(header.driverName);
      in.skipBytes(4); // Reserved
    }
  }

  protected void readFieldDescriptors(DataInputStream in, DBFHeader header) throws IOException {
    byte terminator = in.readByte();
    List<FieldDescriptor> descriptors = new ArrayList<>();
    recordSize = 0;
    while (terminator != 0x0D) {
      // The terminator is not actually a terminator; it is the first byte in the first attribute
      FieldDescriptor descriptor = new FieldDescriptor();
      if ((header.version & 0x7) == 4) {
        // dBase version 7
        descriptor.fieldName = new byte[32];
        descriptor.fieldName[0] = terminator;
        in.read(descriptor.fieldName, 1, descriptor.fieldName.length - 1);
        descriptor.fieldType = in.readByte();
        descriptor.fieldLength = (short) in.readUnsignedByte();
        descriptor.decimalCount = (byte) in.readUnsignedByte();
        in.skipBytes(2); // Reserved.
        descriptor.mdxFlag = in.readByte();
        in.skipBytes(2); // Reserved.
        descriptor.nextAutoIncrementValue = IOUtil.readIntLittleEndian(in);
        in.skipBytes(4); // Reserved.
      } else {
        // dBase older than version 7
        descriptor.fieldName = new byte[11];
        descriptor.fieldName[0] = terminator;
        in.read(descriptor.fieldName, 1, descriptor.fieldName.length - 1);
        descriptor.fieldType = in.readByte();
        in.skipBytes(4);
        descriptor.fieldLength = (short) in.readUnsignedByte();
        descriptor.decimalCount = (byte) in.readUnsignedByte();
        in.skipBytes(13);
        descriptor.mdxFlag = in.readByte();
      }

      recordSize += descriptor.fieldLength;
      descriptors.add(descriptor);
      terminator = in.readByte();
    }
    header.fieldDescriptors = descriptors.toArray(new FieldDescriptor[descriptors.size()]);
  }

  public static final byte EOFMarker = 0x1A;
  public static final byte ValidRecordMarker = ' ';
  public static final byte DeletedRecordMarker = '*';

  public boolean nextKeyValue() throws IOException {
    if (eofReached)
      return false;
    while (true) {
      byte marker = in.readByte();
      iRecord++;
      if (marker == EOFMarker) {
        eofReached = true;
        return false;
      } else if (marker == ValidRecordMarker) {
        // A valid record. Read it.
        in.readFully(this.recordValue);
        this.feature = readFeatureFields(this.recordValue, this.header);
        return true;
      } else if (marker == DeletedRecordMarker) {
        // A deleted record. Skip over it
        in.readFully(this.recordValue);
      } else {
        String filename = dbfPath == null? "DBF" : "'" + dbfPath + "'";
        throw new RuntimeException(String.format("Error parsing %s file. Invalid marker %d", filename, marker));
      }
    }
  }

  private Feature readFeatureFields(byte[] fieldValues, DBFHeader header) {
    int startOffset = 0;
    Object[] values = new Object[1 + header.fieldDescriptors.length];
    values[0] = null; // Set the geometry component to null
    for (int iField = 0; iField < header.fieldDescriptors.length; iField++) {
      FieldDescriptor fieldDescriptor = header.fieldDescriptors[iField];
      int endOffsetInclusive = Math.min(fieldValues.length - 1, startOffset + fieldDescriptor.fieldLength - 1);
      Object fieldValue;
      switch (fieldDescriptor.fieldType) {
        case DBFConstants.TypeString:
          // String
          while (endOffsetInclusive >= startOffset &&
              (fieldValues[endOffsetInclusive] == ' ' || fieldValues[endOffsetInclusive] == 0))
            endOffsetInclusive--;
          fieldValue = startOffset > endOffsetInclusive? null :
              new String(fieldValues, startOffset, endOffsetInclusive - startOffset + 1, StandardCharsets.UTF_8);
          break;
        case DBFConstants.TypeNumeric:
        case DBFConstants.TypeFloat:
          // Skip any leading blanks (or non-digits)
          int adjustedStartOffset = startOffset;
          while (adjustedStartOffset <= endOffsetInclusive &&
              !(fieldValues[adjustedStartOffset] >= '0' && fieldValues[adjustedStartOffset] <= '9') &&
              fieldValues[adjustedStartOffset] != '.' && fieldValues[adjustedStartOffset] != '-')
            adjustedStartOffset++;
          // Check if it is an empty value
          if (adjustedStartOffset > endOffsetInclusive)
            fieldValue = null;
          else {
            String strValue = new String(fieldValues, adjustedStartOffset, endOffsetInclusive - adjustedStartOffset + 1, StandardCharsets.US_ASCII);
            if (fieldDescriptor.decimalCount == 0 && fieldDescriptor.fieldLength < 10) {
              // Return the value as integer
              fieldValue = Integer.parseInt(strValue);
            } else if (fieldDescriptor.decimalCount == 0 && fieldDescriptor.fieldLength < 19) {
              // Return value as long
              fieldValue = Long.parseLong(strValue);
            } else {
//              strValue = strValue.substring(0, strValue.length() - fieldDescriptor.decimalCount) + "."
//                    + strValue.substring(strValue.length() - fieldDescriptor.decimalCount);
              fieldValue = Double.parseDouble(strValue);
            }
          }
          break;
        case DBFConstants.TypeDouble:
          // Double
          double v = IOUtil.readDoubleLittleEndian(fieldValues, startOffset);
          fieldValue = Double.isNaN(v) ? null : v;
          break;
        case DBFConstants.TypeDate:
          // Date - 8 bytes - date stored as a string in the format YYYYMMDD.
          int yyyy = (fieldValues[startOffset] - '0') * 1000 + (fieldValues[startOffset + 1] - '0') * 100 +
              (fieldValues[startOffset + 2] - '0') * 10 + (fieldValues[startOffset + 3] - '0');
          int mm = (fieldValues[startOffset + 4] - '0') * 10 + (fieldValues[startOffset + 5] - '0');
          int dd = (fieldValues[startOffset + 6] - '0') * 10 + (fieldValues[startOffset + 7] - '0');
          GregorianCalendar date = new GregorianCalendar(DBFConstants.UTC);
          date.clear();
          date.set(yyyy, mm - 1, dd);
          fieldValue = new java.sql.Date(date.getTimeInMillis());
          break;
        case DBFConstants.TypeDatetime:
          // 8 bytes - two longs, first for date, second for time.
          // The date is the number of days since  01/01/4713 BC.
          // Time is hours * 3600000L + minutes * 60000L + Seconds * 1000L
          int datepart = IOUtil.readIntLittleEndian(fieldValues, startOffset);
          int timepart = IOUtil.readIntLittleEndian(fieldValues, startOffset + 4);
          if (datepart == 0 && timepart == 0)
            fieldValue = null;
          else {
            GregorianCalendar datetime = new GregorianCalendar(DBFConstants.UTC);
            datetime.clear();
            datetime.setTimeInMillis(datepart * DBFConstants.MillisInOneDay + DBFConstants.DBFEpoch.getTimeInMillis() + timepart);
            fieldValue = new java.sql.Timestamp(datetime.getTimeInMillis());
          }
          break;
        case DBFConstants.TypeBlockNumber:
          // 10 digits representing a .DBT block number. The number is stored as a string,
          // right justified and padded with blanks.
          // Skip leading spaces
          while (startOffset <= endOffsetInclusive && fieldValues[startOffset] == ' ')
            startOffset++;
          fieldValue = startOffset == endOffsetInclusive ? null :
              new String(fieldValues, startOffset, endOffsetInclusive - startOffset + 1, StandardCharsets.UTF_8);
          break;
        case DBFConstants.TypeBoolean:
          // Logical (Boolean) type
          // initialized to 0x20 (space) otherwise T or F
          switch (fieldValues[startOffset]) {
            case ' ': fieldValue = null; break;
            case 'T': fieldValue = true; break;
            case 'F': fieldValue = false; break;
            default: fieldValue = null; break;
          }
          break;
        default:
          throw new RuntimeException(String.format("Field type '%c' not supported", fieldDescriptor.fieldType));
      }
      values[iField + 1] = fieldValue;
      startOffset += fieldDescriptor.fieldLength;
    }
    return new Feature(values, schema,0);
  }

  public IFeature getCurrentValue() {
    return feature;
  }

  public float getProgress() {
    return (float)iRecord / header.numRecords;
  }

  @Override
  public void close() throws IOException {
    if (in != null) {
      in.close();
      in = null;
    }
  }
}
